src/NerdBank.GitVersioning/VersionOracle.cs

```diff
@@ -269,10 +277,16 @@ public IEnumerable<string> BuildMetadataWithCommitId
     public int VersionHeightOffset => this.VersionOptions?.VersionHeightOffsetOrDefault ?? 0;
 
     /// <summary>
-    /// Gets or sets the ref (branch or tag) being built.
+    /// Gets or sets the ref (branch or tag) being built. Just contains a tag if it is known
+    /// that explicitly this tag is build, e.g. in a cloud build context.
     /// </summary>
     public string? BuildingRef { get; protected set; }
 
+    /// <summary>
+    /// Gets or sets a collection of the tags that reference HEAD.
+    /// </summary>
+    public IReadOnlyCollection<string>? BuildingTags { get; protected set; }
```

Comment:
To prevent this property from being outputted in the `CloudBuildAllVars`, you can add the `Ignore` attribute:
```suggestion
    [Ignore]
    public IReadOnlyCollection<string>? BuildingTags { get; protected set; }
```

Reply:
This attribute should also be added to the `VersionOptions` property, since that is now outputted as:
```json
"CloudBuildAllVars": {
  // ...
  "NBGV_VersionOptions": "Nerdbank.GitVersioning.VersionOptions",
  // ...
}
```

Reply:
To fix the output format, the `CloudBuildAllVars` property body can be updated to support converting different values to strings:

```c#
private const string IsoDateTimeFormat = "yyyy'-'MM'-'dd'T'HH':'mm':'ss.FFFFFFFK";

public IDictionary<string, string> CloudBuildAllVars
{
    get
    {
        var variables = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        PropertyInfo[]? properties = this.GetType().GetTypeInfo().GetProperties(BindingFlags.Public | BindingFlags.DeclaredOnly | BindingFlags.Instance);
        foreach (PropertyInfo? property in properties)
        {
            if (property.GetCustomAttribute<IgnoreAttribute>() is null)
            {
                object? propertyValue = property.GetValue(this);
                if (propertyValue is null)
                {
                    continue;
                }

                string value = propertyValue switch
                {
                    IEnumerable<string> values => string.Join(",", values),
                    DateTime dateTime => dateTime.ToString(IsoDateTimeFormat),
                    DateTimeOffset dateTimeOffset => dateTimeOffset.ToString(IsoDateTimeFormat),
                    IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
                    _ => propertyValue.ToString() ?? string.Empty,
                };

                variables.Add($"NBGV_{property.Name}", value);
            }
        }

        return variables;
    }
}
```

This also fixes formatting of dates and other values that use the default culture, so instead of outputting:
```json
"CloudBuildAllVars": {
  // ...
  "NBGV_GitCommitDate": "15-12-2022 21:40:21 +00:00",
  // ...
}
```
This will now output it as (similar to the `GitCommitDate` value is the root of the JSON object):
```json
"CloudBuildAllVars": {
  // ...
  "NBGV_GitCommitDate": "2022-12-15T21:40:21+00:00",
  // ...
}
```

Reply:
Thanks for pointing me in the `[Ignore]` direction, that seems like the right way to exclude it. On the other hand, excluding might not be what we want as I think `BuildingTags` can provide valuable information to a caller/someone who tries to understand what nbgv does why in a specific case.

`VersionOptions ` is a good point, I agree we should probably handle this in the same way as we handle `BuildingTags` (if we keep both; or: `[Ignore]` `VersionOptions` and handle `BuildingTags` differently).

I also think it's preferrable to use ISO date time formats. However, that seems like a breaking change to me. Also, maybe there is a better way to handle all the different possible property types than to switch on every type that comes to our mind? How to hande enumerable types, such as `BuildingTags`, if we want to include these? Comma-separate and recurse? Maybe json serializing each value on its own is a sane and also very simple and well-defined option?

Happy to implement any of these options - I'd be interested in @AArnott's opinion first though :).

Reply:
Adding `[Ignore]` will only remove the `NBGV_BuildingTags` item from the `CloudBuildAllVars` JSON array and not set the environment variable when `cloudBuild.setAllVariables` is enabled; the values will still be available in the JSON output as `BuildingTags` (as shown in your example output). The same is true for `VersionOptions`...

Since environment variable values can only be strings and the current implementation simply uses `ToString()`, it uses the current culture, doesn't support formats and can't serialize complex objects (like string arrays/lists).

The current culture issue can be fixed by using `Convert.ToString(value, CultureInfo.InvariantCulture)` instead of `value.ToString()`, but that could potentially also result in a different output format for dates. Since the current output format is unspecified, I don't really see this as being a breaking change though.

Regarding enumerable types, the default environment variable value separator on Windows is a semicolon (``;`), but a colon (`:`) on other OSes. The above code is using a comma (`,`) to separate the tags, but looking at the [git documentation](https://git-scm.com/docs/git-check-ref-format), both semicolons and commas are allowed in tag names, making the colon the only robust separator (for this specific value).

Getting input from @AArnott before implementing any changes would indeed help and avoid doing unnecessary work 👍🏻

Reply:
I agree that separating the culture concerns from this and creating a second PR makes sense. I also added the suggested `[Ignore]` attribute.

Reply:
I agree with *not* exposing the array in an environment variable before someone needs it.